include/dali/imgcodec/image_format.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_IMAGE_FORMAT_H_
+#define DALI_IMGCODEC_IMAGE_FORMAT_H_
+
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+#include "dali/core/span.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/core/stream.h"
+#include "dali/imgcodec/image_source.h"
+
+namespace dali {
+namespace imgcodec {
+
+struct ImageInfo {
+  TensorShape<> shape;
+  struct {
+    int rotate;
+    bool flip_x;
+    bool flip_y;
+  } orientation;
+};
+
+class ImageParser {
+ public:
+  virtual ~ImageParser() = default;
+  virtual ImageInfo GetInfo(ImageSource *encoded) const = 0;
+  virtual bool CanParse(ImageSource *encoded) const = 0;
+};
+
+class ImageCodec;
+
+class DLL_PUBLIC ImageFormat {
+ public:
+  ImageFormat(const char *name, shared_ptr<ImageParser> parser);
+
+  bool Matches(ImageSource *encoded) const;
+  ImageParser* Parser() const;
+  const std::string& Name() const;
+  span<ImageCodec* const> Codecs() const;
+  void RegisterCodec(std::shared_ptr<ImageCodec> codec, float priority);
+
+ private:
+  std::string name_;
+  std::shared_ptr<ImageParser> parser_;
+  std::multimap<float, std::shared_ptr<ImageCodec>> codecs_;
+  std::vector<ImageCodec*> codec_ptrs_;
+};
+
+class DLL_PUBLIC ImageFormatRegistry {
+ public:
+  /**
+   * @brief Registers a new image format
+   *
+   * @param format The image format to register
+   */
+  void RegisterFormat(std::shared_ptr<ImageFormat> format);
+  /**
+   * @brief Get the format of the image encoded in `image`
+   *
+   * This function tries to parse the image with parsers from all
+   * registered formats and returns the first format that succeeded.
+   *
+   * @param image
+   * @return ImageFormat*
+   */
+  ImageFormat* GetImageFormat(ImageSource *image) const;
```

Comment:
As I understand, we get `ImageSource` from the reader, then we look here for matching `ImageFormat`. We do so by using `CanParse` method of a parser associated with that `ImageFormat`. Is this correct?
And this is done, because `ImageSource` is basically raw data with some small additions.

I think that more often than not, we know the format of a file based on its extension or some other external information (like DL dataset description). Having this, there is no need to dynamically check, which parser matches the `ImageSource` and by extension what is the format. 
I get that in general, maybe we do not now the format in the first place and we need this but I think that based on our use cases this is rather an exceptional situation. Is this assumption correct?

Are there any plans to include something like that?  Or maybe parsers can depend on `ImageSource.SourceInfo` for some additional context.

I am asking because I am a bit afraid that it will not be easy to write those `CanParse` functions and this method of finding a format relies on them. Especially with many formats and matching format being at the end of the list, we might see some performance penalties, if the `CanParse` methods are slow.

Reply:
This will be basically Cut/Paste (with slight modifications) from functions like `CheckIsJPEG`, `CheckIsPNG`, etc which we already have. The biggest difference is that the new functions will work with ImageSource rather than raw memory.

Reply:
Yeah, so the question is, do we want to replicate this approach or not?

Since this whole thing is being redone from scratch this is the time to question assumptions made before, I think.

Reply:
Thanks for the clarification. 
 
I agree that generally there is not much overhead to check this. Maybe I'm just a bit skewed by the issues from video space where the only 100% solid way to check something is to decode it :)